hydromt/models/root.py

```diff
@@ -0,0 +1,66 @@
+"""the new model root class."""
+
+from logging import Logger, getLogger
+from os import PathLike
+from os.path import isdir
+from pathlib import Path
+
+from hydromt.io.path import make_path_abs_and_cross_platform
+from hydromt.typing import ModeLike, ModelMode
+
+logger = getLogger(__name__)
+
+
+class ModelRoot:
+    """A class to handle model roots in a cross platform manner."""
+
+    def __init__(
+        self,
+        path: PathLike,
+        mode: ModeLike = "w",
+        logger: Logger = logger,
+    ):
+        self.mode = mode
+        self.path = path
```

Comment:
I'm not familiar with pydantic and @.setter so just to understand, when calling this line, it's equivalent to calling a eg set_path(path) function but in your case it link to the path setter function?

So in case you want to update the path, how does this work? ModelRoot.path = new_path? And the setter function is re-run?

Why in case of path the setter function returns a Path and for mode nothing?

Reply:
This is not anying to do with Pydantic, this is a function of the `@property` decorator. You are correct that `ModelRoot.path = new_path` should be equivalent to running `ModelRoot.set_path(new_path)`. The mode doesn't return anything because that is a mistake of mine, well caught!